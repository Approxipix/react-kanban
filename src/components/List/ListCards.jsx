import React, { Component, PureComponent } from 'react';
import styled from 'styled-components'
import { Droppable } from 'react-beautiful-dnd';
import Card from '../Card/Card'

const CardLIst = styled.div`
  min-height: 2rem;
  position: relative;
  padding: .5rem;
  flex-grow: 1;
  overflow: auto;
`;

class InnerList extends PureComponent {
  render() {
    const { listId, cardsId, } = this.props;
    return (
      cardsId.map((cardId, index) =>
        <Card
          isDraggingOver={this.props.isDraggingOver}
          key={index}
          index={index}
          listId={listId}
          cardId={cardId}
        />
      )
    )
  }
}

class ListCards extends Component {
  componentDidUpdate = prevProps => {
    const { cardsId } = this.props;
    if (
      cardsId[cardsId.length - 2] ===
      prevProps.cardsId[prevProps.cardsId.length - 1]
    ) {
      this.scrollToBottom();
    }
  };

  scrollToBottom = () => {
    this.listEnd.scrollIntoView();
  };

  render() {
    const { listId, cardsId } = this.props;
    return (
      <Droppable
        type={'task'}
        droppableId={`${listId}`}
      >
        {(provided, { isDraggingOver }) => (
          <CardLIst
            {...provided.droppableProps}
            ref={provided.innerRef}
            isDraggingOver={isDraggingOver}
          >
            <InnerList
              isDraggingOver={isDraggingOver}
              listId={listId}
              cardsId={cardsId}
            />
            {provided.placeholder}
            <div ref={el => this.listEnd = el}/>
          </CardLIst>
        )}
      </Droppable>
    )
  }
}

export default ListCards;
