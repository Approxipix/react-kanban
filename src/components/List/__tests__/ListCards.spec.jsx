import React from 'react';
import configureStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import toJson from "enzyme-to-json";
import ListCards from '../ListCards';

const mockStore = configureStore([]);
jest.mock('react-beautiful-dnd', () => ({
  Droppable: ({ children }) => children({
    draggableProps: {
      style: {},
    },
    innerRef: jest.fn(),
  }, {}),
  Draggable: ({ children }) => children({
    draggableProps: {
      style: {},
    },
    innerRef: jest.fn(),
  }, {}),
  DragDropContext: ({ children }) => children,
}));

describe('<ListCards>', () => {
  let store;
  let component;

  beforeEach(() => {
    store = mockStore({
      listReducer: {
        ListID: {
          _listId: 'ListID',
          title: 'List Title',
          cards: []
        }
      },
    });
    const props = {
      listId: 'ListID',
      cardsId: ['Card1ID', 'Card2ID', 'Card3ID']
    };

    component = shallow(<ListCards store={store} {...props} />).dive();
  });

  it('should render with given state from Redux store', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render a Card component for each card id', () => {
    expect(component.find('InnerList').dive().length).toEqual(3)
  });
});
