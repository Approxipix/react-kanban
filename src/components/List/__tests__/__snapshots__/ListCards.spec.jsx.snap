// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListCards> should render with given state from Redux store 1`] = `
<styled.div>
  <InnerList
    cardsId={
      Array [
        "Card1ID",
        "Card2ID",
        "Card3ID",
      ]
    }
  />
  <div />
</styled.div>
`;
